core/pbcc_subscribe_event_engine_events.c

```diff
@@ -0,0 +1,333 @@
+        SUBSCRIBE_EE_EVENT_SUBSCRIPTION_CHANGED;
+    return pbcc_subscribe_ee_event_alloc_(ee,
+                                          type,
+                                          channels,
+                                          channel_groups,
+                                          NULL,
+                                          NULL);
+}
+
+pbcc_ee_event_t* pbcc_subscription_restored_event_alloc(
+    pbcc_subscribe_ee_t*            ee,
+    char**                          channels,
+    char**                          channel_groups,
+    const pubnub_subscribe_cursor_t cursor)
+{
+    if (NULL != channels && 0 == strlen(*channels)) {
+        free(*channels);
+        *channels = NULL;
+    }
+    if (NULL != channel_groups && 0 == strlen(*channel_groups)) {
+        free(*channel_groups);
+        *channel_groups = NULL;
+    }
+
+    const pbcc_subscribe_ee_event type =
+        SUBSCRIBE_EE_EVENT_SUBSCRIPTION_RESTORED;
+    return pbcc_subscribe_ee_event_alloc_(ee,
+                                          type,
+                                          channels,
+                                          channel_groups,
+                                          &cursor,
+                                          NULL);
+}
+
+pbcc_ee_event_t* pbcc_handshake_success_event_alloc(
+    pbcc_subscribe_ee_t*            ee,
+    const pubnub_subscribe_cursor_t cursor)
+{
+    const pbcc_subscribe_ee_event type = SUBSCRIBE_EE_EVENT_HANDSHAKE_SUCCESS;
+    return pbcc_subscribe_ee_event_alloc_(ee, type, NULL, NULL, &cursor, NULL);
+}
+
+pbcc_ee_event_t* pbcc_handshake_failure_event_alloc(
+    pbcc_subscribe_ee_t*  ee,
+    const enum pubnub_res reason)
+{
+    const pbcc_subscribe_ee_event type = SUBSCRIBE_EE_EVENT_HANDSHAKE_FAILURE;
+    return pbcc_subscribe_ee_event_alloc_(ee, type, NULL, NULL, NULL, &reason);
+}
+
+pbcc_ee_event_t* pbcc_receive_success_event_alloc(
+    pbcc_subscribe_ee_t*            ee,
+    const pubnub_subscribe_cursor_t cursor)
+{
+    const pbcc_subscribe_ee_event type = SUBSCRIBE_EE_EVENT_RECEIVE_SUCCESS;
+    return pbcc_subscribe_ee_event_alloc_(ee, type, NULL, NULL, &cursor, NULL);
+}
+
+pbcc_ee_event_t* pbcc_receive_failure_event_alloc(
+    pbcc_subscribe_ee_t*  ee,
+    const enum pubnub_res reason)
+{
+    const pbcc_subscribe_ee_event type = SUBSCRIBE_EE_EVENT_RECEIVE_FAILURE;
+    return pbcc_subscribe_ee_event_alloc_(ee, type, NULL, NULL, NULL, &reason);
+}
+
+pbcc_ee_event_t* pbcc_disconnect_event_alloc(pbcc_subscribe_ee_t* ee)
+{
+    const pbcc_subscribe_ee_event type = SUBSCRIBE_EE_EVENT_DISCONNECT;
+    return pbcc_subscribe_ee_event_alloc_(ee, type, NULL, NULL, NULL, NULL);
+}
+
+pbcc_ee_event_t* pbcc_reconnect_event_alloc(
+    pbcc_subscribe_ee_t*            ee,
+    const pubnub_subscribe_cursor_t cursor)
+{
+    const pbcc_subscribe_ee_event type = SUBSCRIBE_EE_EVENT_RECONNECT;
+    return pbcc_subscribe_ee_event_alloc_(ee, type, NULL, NULL, &cursor, NULL);
```

Comment:
Conclusion, not a review note:
When I see the C code I would really love to see builders... 
There is a lot of `NULL` statements that means nothing during the code review. 

![image](https://github.com/user-attachments/assets/dfdd70e0-458a-4678-b4d1-136c07ed846a)

Reply:
IDE will show docs for those arguments, so it will have a bit more meaning :)